cirq/contrib/shannon_decomposition/three_qubit_decomposition.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+from cirq.linalg import is_unitary
+from cirq.ops import YY
+from cirq.protocols import unitary
+
+_YY = unitary(YY)
+
+
+def _to_special(u: np.ndarray) -> np.ndarray:
+    """Converts a unitary matrix to a special unitary matrix.
+
+    All unitary matrices u have |det(u)| = 1.
+    Also for all d dimensional unitary matrix u, and scalar s:
+        det(u * s) = det(u) * s^(d)
+    To find a special unitary matrix from u:
+        u * det(u)^{-1/d}
+
+    Args:
+        u: the unitary matrix
+    Returns:
+        the special unitary matrix
+    """
+    return u * (np.linalg.det(u)**(-1 / len(u)))
+
+
+def _gamma(u: np.ndarray) -> np.ndarray:
+    """Gamma function to convert u to the magic basis.
+
+    See Definition IV.1 in Shende et al. "Minimal Universal Two-Qubit CNOT-based
+     Circuits." https://arxiv.org/abs/quant-ph/0308033
+
+    Args:
+        u: a member of SU(4)
+    Returns:
+        u @ yy @ u.T @ yy, where yy = Y ⊗ Y
+    """
+
+    return u @ _YY @ u.T @ _YY
+
+
+def _is_three_cnot_two_qubit_unitary(u: np.ndarray) -> bool:
+    """Returns true if U requires 3 CNOT/CZ gates.
+
+    See Proposition III.1, III.2, III.3 in Shende et al. “Recognizing Small-
+    Circuit Structure in Two-Qubit Operators and Timing Hamiltonians to Compute
+    Controlled-Not Gates”.  https://arxiv.org/abs/quant-ph/0308045
+
+    Args:
+        u: a two-qubit unitary
+    Returns:
+        the number of two-qubit gates required to implement the unitary
+    """
+    assert np.shape(u) == (4, 4)
+    assert is_unitary(u)
+
+    poly = np.poly(_gamma(_to_special(u)))
```

Comment:
By the last part of proposition III.3 in ["Recognizing..."](https://arxiv.org/pdf/quant-ph/0308045.pdf), if your goal is to check whether u requires 3 CNOTs then you don't need to compute the full characteristic polynomial of gamma. The reason is that coefficients 0, 2 and 4 are guaranteed to be real and coefficients 3 and 1 are the trace and its conjugate. Thus, we only need to check that the trace is real which is simpler.

```python
g = _gamma(_to_special(u))
return not np.allclose(np.trace(g).imag, 0, atol=tolerance)
```

Reply:
(This comment becomes moot if you choose to make this a function that computes the number of CNOTs required.)

Reply:
Actually, even in this more general case we can avoid computing the characteristic polynomial, because the symmetry of the polynomial used in Proposition III.3 remains valid and so we only need to know the coefficients by x^3 and x^2. Specifically, I think we can do something like this:

```python
def number_of_cnots_required(u: np.ndarray, tolerance: float = 1e-6) -> int:
    g = _gamma(_to_special(u))
    a3 = np.trace(g)
    if np.abs(a3 - 4) < tolerance or np.abs(a3 + 4) < tolerance:
        return 0
    a2 = (a3*a3 - np.trace(g @ g)) / 2
    if np.abs(a3) < tolerance and np.abs(a2 - 2) < tolerance:
        return 1
    if np.abs(a3.imag) < tolerance:
        return 2
    return 3
```

The first condition comes from Proposition III.1 and Pascal triangle. Note that we don't need to check whether a2 = 6 (why?). The second condition comes from Proposition III.2 and simple algebra. The final condition is the existing code.

The formula for a2 is the second iteration of [Fadeev-LaVerrier](https://en.wikipedia.org/wiki/Faddeev%E2%80%93LeVerrier_algorithm).

Reply:
This is great - it is probably more efficient than my naive https://github.com/quantumlib/Cirq/pull/2873/files#diff-4a594a63b684a45eefea705168a2777bR193 following the polynomial calculation - assuming that np.poly is more expensive than calculating g @ g. 

Thank you for pointing me to the recursive Fadeev-LaVerrier formula. 

Not exactly sure why we don't need to check whether a2 = 6. We know from `g` being from SU(4) and that the product of the eigenvalues should be 1 that  a4=a0=1 and a1 = a3 = tr(g) as described in the proof of Proposition III.3. Fadeev-LaVerrier requires g^2 to be calculated, so it can't be a simple function of a3...but it sounds like you have a reason to believe that from a3 = ±4 it follows that a2 = 6. I would need to find a special unitary with char.polynomial [1,4,a3!=6,4,1] to prove you wrong, but I'm going to ask for your proof instead :)

This looks good - not that it matters for the calculations, but to stay closer to the Fadeev-LaVerrier formulae, `a3 = - np.trace(g)`. Also `np.abs(a3)-4 < tolerance` is a further simplification of the first condition.

Reply:
(I misspelled the name in my comment: should be "Le" in LeVerrier. Sorry.)

Proof that a3 = ±4 implies a2 = 6: Consider the four eigenvalues of g. They are all numbers on the unit circle in the complex plane. The only way(*) for a3, i.e. the negative sum of the eigenvalues to be +4 (respectively, -4) is if they are all +1 (respectively, -1). So the condition a3 = ±4 fixes all eigenvalues and hence it also fixes the characteristic polynomial. The matrix with eigenvalues +1, +1, +1, +1 (respectively, -1, -1, -1, -1) has characteristic polynomial (x-1)^4 (respectively, (x+1)^4) which is x^4 + 4x^3 + 6x^2 + 4x + 1 (respectively, x^4 - 4x^3 + 6x^2 - 4x + 1). Thus, a2 = 6. Done.

Regarding the efficiency of np.poly. My guess is that it uses Fadeev-LeVerrier, so it's probably quite fast, but will still do more work than we need because it is unaware of highly constrained nature of our matrix. In particular, for the last coefficient it will end up computing the determinant as it doesn't know that our matrix is in SU(4)!

I agree `a3 = -np.trace(g)` is better. Thanks for catching this!

I don't think we can simplify `np.abs(a3 - 4) < tolerance or np.abs(a3 + 4) < tolerance` to `np.abs(a3)-4 < tolerance`, because we don't know that a3 is real! Counterexample is `np.diag(1j, 1j, 1j, 1j)`.

---
(*) Formally: let the real parts of the eigenvalues be 1-a, 1-b, 1-c, 1-b. Note that a, b, c, d are real numbers in [0, 2] (or else the eigenvalues are not on the unit circle). The sum of the eigenvalues is 4-a-b-c-d which equals 4 only when a=b=c=d=0.

Reply:
oh such a simple proof, nice :) I should have just thought a bit harder, thank you! 
Sounds good, I'll do the modifications!